sdk/python/lib/pulumi/_utils.py

```diff
@@ -0,0 +1,57 @@
+# Copyright 2016-2020, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import typing
+
+
+# Empty function definitions.
+
+def _empty():
+    ...
+
+def _empty_doc():
+    """Empty function docstring."""
+    ...
+
+_empty_lambda = lambda: None
+
+_empty_lambda_doc = lambda: None
+_empty_lambda_doc.__doc__ = """Empty lambda docstring."""
+
+
+def _consts(fn: typing.Callable) -> tuple:
+    """
+    Returns a tuple of the function's constants excluding the docstring.
+    """
+    return tuple(x for x in fn.__code__.co_consts if x != fn.__doc__)
+
+
+# Precompute constants for each of the empty functions.
+_consts_empty_ = _consts(_empty)
+_consts_empty_doc = _consts(_empty_doc)
+_consts_empty_lambda = _consts(_empty_lambda)
+_consts_empty_lambda_doc = _consts(_empty_lambda_doc)
+
+
+def is_empty_function(fn: typing.Callable) -> bool:
+    """
+    Returns true if the function is empty.
+    """
+    consts = _consts(fn)
+    return (
+        (fn.__code__.co_code == _empty.__code__.co_code and consts == _consts_empty_) or
+        (fn.__code__.co_code == _empty_doc.__code__.co_code and consts == _consts_empty_doc) or
+        (fn.__code__.co_code == _empty_lambda.__code__.co_code and consts == _consts_empty_lambda) or
+        (fn.__code__.co_code == _empty_lambda_doc.__code__.co_code and consts == _consts_empty_lambda_doc)
+    )
```

Comment:
You could also check if the function is equivalent to returning `None` at the Bytecode level.

```suggestion
def is_empty_function(fn: typing.Callable) -> bool:
    """
    Returns true if the function is empty.
    """
    import dis
    fst, snd, *rest = [i for i in dis.Bytecode(fn)] 
    is_load_none = fst.opname == "LOAD_CONST" and fst.argval == None 
    is_return = snd.opname == "RETURN_VALUE" 
    return is_load_none and is_return and not rest 
```

Reply:
Thanks for the suggestion, @KentShikama! I tried both approaches and found the original approach was faster in microbenchmarks:

```python
import typing
import dis

def _empty():
    ...

def _empty_doc():
    """Empty function docstring."""
    ...

_empty_lambda = lambda: None

_empty_lambda_doc = lambda: None
_empty_lambda_doc.__doc__ = """Empty lambda docstring."""


def _consts(fn: typing.Callable) -> tuple:
    """
    Returns a tuple of the function's constants excluding the docstring.
    """
    return tuple(x for x in fn.__code__.co_consts if x != fn.__doc__)


# Precompute constants for each of the empty functions.
_consts_empty = _consts(_empty)
_consts_empty_doc = _consts(_empty_doc)
_consts_empty_lambda = _consts(_empty_lambda)
_consts_empty_lambda_doc = _consts(_empty_lambda_doc)


def is_empty_function(fn: typing.Callable) -> bool:
    """
    Returns true if the function is empty.
    """
    consts = _consts(fn)
    return (
        (fn.__code__.co_code == _empty.__code__.co_code and consts == _consts_empty) or
        (fn.__code__.co_code == _empty_doc.__code__.co_code and consts == _consts_empty_doc) or
        (fn.__code__.co_code == _empty_lambda.__code__.co_code and consts == _consts_empty_lambda) or
        (fn.__code__.co_code == _empty_lambda_doc.__code__.co_code and consts == _consts_empty_lambda_doc)
    )

def is_empty_function_dis(fn: typing.Callable) -> bool:
    """
    Returns true if the function is empty.
    """
    fst, snd, *rest = [i for i in dis.Bytecode(fn)]
    is_load_none = fst.opname == "LOAD_CONST" and fst.argval == None
    is_return = snd.opname == "RETURN_VALUE"
    return is_load_none and is_return and not rest


def empty():
    pass

def not_empty():
    return 42

if __name__ == '__main__':
    import timeit
    print("Original empty")
    print(timeit.repeat("is_empty_function(empty)",
                        setup="from __main__ import empty, is_empty_function",
                        repeat=3))
    print("Original non-empty")
    print(timeit.repeat("is_empty_function(not_empty)",
                        setup="from __main__ import not_empty, is_empty_function",
                        repeat=3))

    print("Suggestion empty")
    print(timeit.repeat("is_empty_function_dis(empty)",
                        setup="from __main__ import empty, is_empty_function_dis",
                        repeat=3))
    print("Suggestion non-empty")
    print(timeit.repeat("is_empty_function_dis(not_empty)",
                        setup="from __main__ import not_empty, is_empty_function_dis",
                        repeat=3))
```

Results (Python 3.8.5):

```
Original empty
[0.862670518, 0.8678176559999999, 0.8616144619999997]
Original non-empty
[1.2970095930000003, 1.301474218, 1.3091111590000004]
Suggestion empty
[7.6551781210000005, 7.600018379999998, 7.587793077000001]
Suggestion non-empty
[7.4596535589999995, 7.886800770000001, 7.810590937000001]
```